### PR TITLE
Added support for Coverity On Demand badges.

### PR DESCRIPTION
**Pull request by jfitzpatrick99**

### CODE REVIEW

**Comment by jvantuyl** [comment]
+1

**Comment by eelgheez** [comment]
+1

**Comment by ascudder** [comment]
I'd like to see this, too!

**Comment by jbyler** [comment]
+1

**Comment by mtaylor91** [comment]
+1

File: server.js [hunk 1]
```diff
@@ -684,6 +684,33 @@ cache(function(data, match, sendBadge, request) {
...
 
+// Coverity Code Advisor On Demand integration
+camp.route(/^\/coverity\/ondemand\/(.+)\/(.+)\.(svg|png|gif|jpg|json)$/,
+cache(function(data, match, sendBadge, request) {
+  var badgeType = match[1]; // streams or jobs
+  var badgeTypeId = match[2]; // streamId or jobId
+  var format = match[3];
+  var url = 'https://api.ondemand.coverity.com/' + badgeType + '/' + badgeTypeId + '/badge';
+//  var url = 'https://api-stage01.caas.coverity.com/' + badgeType + '/' + badgeTypeId + '/badge';
+//  var url = 'http://localhost:21020/' + badgeType + '/' + badgeTypeId + '/badge';
```
Comment (espadrine) [inline server.js]: Why are they in comments? Is there one that is more valid?

File: server.js [hunk 2]
```diff
@@ -684,6 +684,33 @@ cache(function(data, match, sendBadge, request) {
...
+  console.log(url)
```
Comment (espadrine) [inline server.js]: In the final code, there shouldn't be a log here.

File: server.js [hunk 3]
```diff
@@ -684,6 +684,33 @@ cache(function(data, match, sendBadge, request) {
   });
 }));
 
+// Coverity Code Advisor On Demand integration
+camp.route(/^\/coverity\/ondemand\/(.+)\/(.+)\.(svg|png|gif|jpg|json)$/,
+cache(function(data, match, sendBadge, request) {
+  var badgeType = match[1]; // streams or jobs
+  var badgeTypeId = match[2]; // streamId or jobId
```
Comment (espadrine) [inline server.js]: Could you provide working examples here? They are hugely useful for debugging purposes.

File: server.js [hunk 4]
```diff
@@ -684,6 +684,33 @@ cache(function(data, match, sendBadge, request) {
...
+  var badgeData = getBadgeData('coverity', data);
+  request(url, function(err, res, buffer) {
+    if (err != null) {
+      badgeData.text[1] = 'inaccessible';
+      sendBadge(format, badgeData);
+      return;
+    }
+    try {
+      badgeData = JSON.parse(buffer);
+      sendBadge(format, badgeData);
```
Comment (espadrine) [inline server.js]: Are you sure the format for the JSON we receive is the same as the one we use internally? It is pretty specific.

**Comment by espadrine** [comment]
Thanks for your contribution!

File: server.js (hunk 1 above)
Comment (jvantuyl) [inline server.js]: I think these are left over from testing (a stage environment and local testing environment).  They can almost certainly be removed.

File: server.js (hunk 3 above)
Comment (jfitzpatrick99) [inline server.js]: Coverity On Demand jobs expire after 30 days which makes including a working example that isn't stale difficult. I have just pushed a change that provides an example of a small project that can be submitted for analysis and tested against.

File: server.js (hunk 4 above)
Comment (jfitzpatrick99) [inline server.js]: I am responsible for maintaining the json that is sent in response to a "badge" request so I have made it conform to the format used by shields. If the shields json badge format changes I will just update the code on our end to track the change.

File: server.js (hunk 3 above)
Comment (espadrine) [inline server.js]: I see. Could we have a static badge in `try.html` to show an example of what it would look like?

File: try.html [hunk 5]
```diff
@@ -172,6 +172,10 @@ <h3 id="build"> Build </h3>
     <td><img src='/coverity/scan/3997.svg' alt=''/></td>
     <td><code>https://img.shields.io/coverity/scan/3997.svg</code></td>
   </tr>
+  <tr><th> Coverity Code Advisor On Demand: </th>
+    <td><img src='/coverity/ondemand/streams/1234.svg' alt=''/></td>
+    <td><code>https://img.shields.io/coverity/ondemand/streams/1234.svg</code></td>
```
Comment (espadrine) [inline try.html]: This being informational, we can probably put `<code>https://img.shields.io/coverity/ondemand/streams/STREAM.svg</code>`.

We can also add another custom badge to show the same thing for jobs.

**Comment by jfitzpatrick99** [comment]
Are there any changes you would like me to make? Is it possible to get this pull request accepted?

**Comment by espadrine** [comment]
Thanks! The server is updated. I'll update the website to show the new badges in 24h. You can perform a few tests in the meantime.

**Comment by jfitzpatrick99** [comment]
Thanks!

FYI, we will be doing a production push mid next-week at which time our new badges feature will be available. It is up to you if you would like to put off a production deploy on your side until we start servicing badge requests.

**Pull request merged**